In [147]:
import numpy as np
import pandas as pd
from scipy import stats

# Домашняя работа №6

**№1.**Даны значения величины заработной платы заемщиков банка (`salary`) и значения их поведенческого кредитного скоринга (`scoring`):
```
salary = [35, 45, 190, 200, 40, 70, 54, 150, 120, 110]
scoring = [401, 574, 874, 919, 459, 739, 653, 902, 746, 832]
```
1. Найдите ковариацию этих двух величин: сначала без использования специальных функций, а затем с помощью функции `numpy.cov`. Полученные значения должны быть равны.
2. Найдите коэффициент корреляции Пирсона: с помощью ковариации и средних квадратических отклонений двух признаков, а затем с помощью специальных функций из библиотек `numpy` и `pandas`.


**Решение пункта 1.**
Пусть $X = [35, 45, 190, 200, 40, 70, 54, 150, 120, 110]$, $Y = [401, 574, 874, 919, 459, 739, 653, 902, 746, 832]$.

Найдем коэффициент ковариации $K_{XY} = M(X-MX)(Y-MY) = MXY - MXMY$. Несмещенная оценка ковариации по выборке:

$$\sigma_{XY} = \dfrac{1}{n - 1} \displaystyle\sum_{i = 1}^n (x_i - \overline{X}) \cdot (y_i - \overline{Y})$$.

$\overline{X} = \dfrac{1}{10}\displaystyle\sum_{i = 1}^{10}x_i = \dfrac{1014}{10} = 101.4$,
$\overline{y} = \dfrac{1}{10}\displaystyle\sum_{i = 1}^{10}y_i = \dfrac{7099}{10} = 709.9$,

$X - \overline{X} = [-66.4, -56.4,  88.6,  98.6, -61.4, -31.4, -47.4,  48.6,  18.6, 8.6]$,
$Y - \overline{Y} = [-308.9, -135.9,  164.1,  209.1, -250.9,   29.1,  -56.9,  192.1, 36.1,  122.1]$,
$(X - \overline{X}) \cdot (Y - \overline{Y}) = [20510.96,  7664.76, 14539.26, 20617.26, 15405.26,  -913.74, 2697.06,  9336.06,   671.46,  1050.06]$,

$\sigma_{XY} = \dfrac{1}{9} \displaystyle\sum_{i = 1}^{10} (x_i - \overline{X}) \cdot (y_i - \overline{Y}) = \dfrac{1}{9}\cdot91578.40000000001 = 10175.37777777778$.

**Ответ:** $K_{XY} = 10175.37777777778$.

In [148]:
# Без использования специальных функций

X = np.array([35, 45, 190, 200, 40, 70, 54, 150, 120, 110])
Y = np.array([401, 574, 874, 919, 459, 739, 653, 902, 746, 832])

MX = X.mean()
MY = Y.mean()

X_MX = X - MX
Y_MY = Y - MY

Z = X_MX  * Y_MY
K = Z.sum()/9

print(f'Коэффициент ковариации X и Y: {K}')

Коэффициент ковариации X и Y: 10175.37777777778


In [149]:
# C использованием встроенных функций
X = np.array([35, 45, 190, 200, 40, 70, 54, 150, 120, 110])
Y = np.array([401, 574, 874, 919, 459, 739, 653, 902, 746, 832])

K = np.cov(X,Y)
print(f'Матрица ковариации:\n {K}\n')
print(f'Коэффициент ковариации X и Y: {K[0,1]}')

Матрица ковариации:
 [[ 3882.93333333 10175.37777778]
 [10175.37777778 33854.32222222]]

Коэффициент ковариации X и Y: 10175.377777777776


**Решение пункта 2.**
Пусть $X = [35, 45, 190, 200, 40, 70, 54, 150, 120, 110]$, $Y = [401, 574, 874, 919, 459, 739, 653, 902, 746, 832]$.

Найдем коэффициент корреляции $r_{XY} = \frac{K_{XY}}{\sqrt{DX\cdot DY}}$. Оценка коэффициента корреляции по выборке:

$$r_{XY} = \dfrac{\sigma_{XY}}{\sigma_X \cdot \sigma_Y}$$.

$D_X = \frac{1}{n-1}\Sigma_{i=1}^{n}(x_i - \overline{x})^2 = 3882.9333333333334$,
$\sigma_X = \sqrt{D_{X}} = 62.31318747531162$,

$D_Y = \frac{1}{n-1}\Sigma_{i=1}^{n}(y_i - \overline{y})^2 = 33854.32222222223$,
$\sigma_Y = \sqrt{D_{Y}} = 183.99544076477068$,


$r_{XY} = \dfrac{10175.377777777776}{62.31318747531162 \cdot 183.99544076477068} = 0.887490092073916$.

**Ответ:** $r_{XY} = 0.887490092073916$.

In [150]:
# Без использования специальных функций

X = np.array([35, 45, 190, 200, 40, 70, 54, 150, 120, 110])
Y = np.array([401, 574, 874, 919, 459, 739, 653, 902, 746, 832])

sX = X.std(ddof=1)
sY = Y.std(ddof=1)

sXY = np.cov(X,Y)[0,1]

r = sXY / (sX * sY)
print(f'Коэффициент корреляции X и Y: {r}')

Коэффициент корреляции X и Y: 0.887490092073916


In [151]:
# C использованием встроенных функций numpy
X = np.array([35, 45, 190, 200, 40, 70, 54, 150, 120, 110])
Y = np.array([401, 574, 874, 919, 459, 739, 653, 902, 746, 832])

r = np.corrcoef(X,Y)
print(f'Матрица корреляции:\n {r}\n')
print(f'Коэффициент корреляции X и Y: {r[0,1]}')

Матрица корреляции:
 [[1.         0.88749009]
 [0.88749009 1.        ]]

Коэффициент корреляции X и Y: 0.8874900920739162


In [152]:
# C использованием встроенных функций pandas
X = [35, 45, 190, 200, 40, 70, 54, 150, 120, 110]
Y = [401, 574, 874, 919, 459, 739, 653, 902, 746, 832]

df = pd.DataFrame({'salary': X, 'scoring': Y})
R = df[['salary', 'scoring']].corr()

print(f'Матрица корреляции:\n {R}\n')
print(f'Коэффициент корреляции X и Y: {R.values[0,1]}')

Матрица корреляции:
           salary  scoring
salary   1.00000  0.88749
scoring  0.88749  1.00000

Коэффициент корреляции X и Y: 0.8874900920739162


**№2.** Измерены значения IQ выборки студентов, обучающихся в местных технических вузах:
```
131, 125, 115, 122, 131, 115, 107, 99, 125, 111
```
Известно, что в генеральной совокупности IQ распределен нормально. Найдите доверительный интервал для математического ожидания с надежностью 0.95.

Дано: $X = [131, 125, 115, 122, 131, 115, 107, 99, 125, 111]$, $X\sim N(a, \sigma)$, $\gamma = 0.95$.

**1 способ (ручной)**. По условию генеральная дисперсия не задана!

1) $\overline{X} = \frac{1}{n}\Sigma_{i=1}^{n}x_i = \frac{1}{10}\Sigma_{i=1}^{10}x_i = 118.1$.

2) $\sigma_X = \sqrt{\frac{1}{n-1}\Sigma_{i=1}^{n}(x_i-\overline{x})^2} = \sqrt{111.2} = 10.55$.

3) Так как $\gamma = 0.95$, $n = 10$, то по таблице значений $t_{\gamma} = 2.26$.

4) $\delta = \frac{t_{\gamma}\cdot \sigma_X}{\sqrt{n}} = \frac{2.26\cdot 10.55}{\sqrt{10}} = 7.54$.

3) Доверительный интервал: $(\overline{X} - \delta; \overline{X} + \delta) = (118.1 - 7.54; 118.1 + 7.54) = (110.55; 125.64)$.

**2 способ (с помощью Python)**.

In [153]:
X = np.array([131,125,115,122,131,115,107,99,125,111])
n = X.shape[0]
mean = X.mean()
std = X.std(ddof=1)

print(f'Выборочное среднее: {mean}')
print(f'Выборочное среднеквадратическое отклонение: {std}')
gamma = 0.95 
alpha = 1 - gamma

t1 = stats.t.ppf(alpha / 2, df = n - 1)
t2 = stats.t.ppf(1 - alpha / 2, df = n - 1)
print(f't1 = {t1}, t2 = {t2} \n')

(mean + t1 * std / np.sqrt(n), mean + t2 * std / np.sqrt(n))

print(f'Доверительный интервал: {(mean + t1 * std / np.sqrt(n), mean + t2 * std / np.sqrt(n))}')

Выборочное среднее: 118.1
Выборочное среднеквадратическое отклонение: 10.54566788359614
t1 = -2.2621571627409915, t2 = 2.2621571627409915 

Доверительный интервал: (110.55608365158724, 125.64391634841274)


**№3.** Известно, что рост футболистов в сборной распределен нормально с известной дисперсией 25. На выборке объёма 27 выборочное среднее составило 174.2. Найдите доверительный интервал для математического ожидания с надежностью 0.95.

**1 способ (ручной)**. По условию генеральная дисперсия задана!

По условию задачи: $X\sim N(a, 5)$, $\gamma = 0.95$, $\overline{x_B} = 174.2$, $n = 27$. Найдем доверительный интервал.

1) Так как $2\cdot \Phi(t_{\gamma}) = \gamma$, то $\Phi(t_{\gamma}) = \frac{\gamma}{2} = \frac{0.95}{2} = 0.475$. Отсюда по таблице значений функции Лапласа $t_{\gamma} = 1.96$ -- коэффициент доверия.

2) Так как $\delta = \frac{t_{\gamma}\cdot \sigma}{\sqrt{n}}$, то $\delta = \frac{1.96\cdot 5}{\sqrt{27}} = 1.89$ -- точность.

3) Доверительный интервал: $(\overline{x_B} - \delta; \overline{x_B} + \delta) = (174.2 - 1.89; 174.2 + 1.89) = (172.31; 176.09)$.

**2 способ (с помощью Python)**.

In [154]:
# данные из условия
n = 27
mean = 174.2
std = 5
gamma = 0.95 # надежность (уровень доверия)

alpha = 1 - gamma
#коэффициент доверия
t1 = stats.norm.ppf(alpha / 2)
t2 = stats.norm.ppf(1 - alpha / 2)

#Доверительный интервал
(mean + t1 * std / np.sqrt(n), mean + t2 * std / np.sqrt(n))

print(f'Доверительный интервал: {(mean + t1 * std / np.sqrt(n), mean + t2 * std / np.sqrt(n))}')

Доверительный интервал: (172.3140237765397, 176.08597622346028)
